File: scripts/build_site.py
#!/usr/bin/env python3
"""
PipeFish Labs - Static Site Build

Assembles the publicly servable website into ./dist so that it can be uploaded
as Cloudflare Workers static assets.

This uses a strict allowlist. Only file types that belong on a public web server
are copied, so repository source (Python, Terraform, Rego, SQL, Helm charts,
Postman collections, tests, CI config) can never be published by accident.
"""

from __future__ import annotations

import argparse
import shutil
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent

# File extensions that are safe to publish.
ALLOWED_SUFFIXES = frozenset(
    {
        ".html",
        ".css",
        ".js",
        ".mjs",
        ".png",
        ".jpg",
        ".jpeg",
        ".gif",
        ".svg",
        ".webp",
        ".avif",
        ".ico",
        ".mp4",
        ".webm",
        ".woff",
        ".woff2",
        ".txt",
        ".xml",
        ".webmanifest",
        ".pdf",
    }
)

# Directories never walked, regardless of extension.
EXCLUDED_DIRS = frozenset(
    {
        ".git",
        ".github",
        ".githooks",
        ".wrangler",
        "__pycache__",
        "node_modules",
        "dist",
        "ansible",
        "bots",
        "charts",
        "db",
        "docs",
        "examples",
        "infra",
        "linkedin",
        "mcp",
        "postman",
        "sales",
        "scripts",
        "security",
        "src",
        "terraform",
        "tests",
    }
)

# Extensionless or otherwise non-allowlisted files that must still be published,
# as paths relative to the repository root.
EXTRA_FILES = (
    "CNAME",
    "_headers",
    "_redirects",
    "admin/config.yml",
    "sdk/openapi.yaml",
    "sdk/asyncapi.yaml",
)

# Cloudflare rejects any single static asset larger than this.
MAX_ASSET_BYTES = 25 * 1024 * 1024


def is_publishable(path: Path) -> bool:
    return path.suffix.lower() in ALLOWED_SUFFIXES


def iter_site_files(root: Path, out_dir: Path):
    """Yield repo-relative paths of every allowlisted file under root."""
    for path in sorted(root.rglob("*")):
        if not path.is_file():
            continue
        # Never pick up a previous build, which would nest output inside itself.
        if out_dir == path or out_dir in path.parents:
            continue
        relative = path.relative_to(root)
        if any(part in EXCLUDED_DIRS for part in relative.parts[:-1]):
            continue
        if not is_publishable(path):
            continue
        yield relative


def collect(root: Path, out_dir: Path) -> tuple[list[Path], list[str]]:
    files = list(iter_site_files(root, out_dir))
    seen = set(files)
    missing: list[str] = []
    for name in EXTRA_FILES:
        relative = Path(name)
        if not (root / relative).is_file():
            missing.append(name)
            continue
        if relative not in seen:
            files.append(relative)
            seen.add(relative)
    return sorted(files), missing


def build(root: Path, out_dir: Path) -> int:
    # rmtree below is destructive, so refuse anything that could be a real tree.
    if out_dir == root or out_dir in root.parents:
        print(
            f"error: refusing to use {out_dir} as the output directory",
            file=sys.stderr,
        )
        return 1

    files, missing = collect(root, out_dir)

    # _headers carries the site's Content-Security-Policy and HSTS. Shipping
    # without it would silently publish a site with no security headers, so a
    # missing entry fails the build instead of warning.
    if missing:
        for name in missing:
            print(f"error: required file is missing: {name}", file=sys.stderr)
        return 1

    if not files:
        print("error: no publishable files found", file=sys.stderr)
        return 1

    if out_dir.exists():
        shutil.rmtree(out_dir)
    out_dir.mkdir(parents=True)

    total_bytes = 0
    oversized: list[tuple[Path, int]] = []

    for relative in files:
        source = root / relative
        size = source.stat().st_size
        total_bytes += size
        if size > MAX_ASSET_BYTES:
            oversized.append((relative, size))
        destination = out_dir / relative
        destination.parent.mkdir(parents=True, exist_ok=True)
        shutil.copy2(source, destination)

    if oversized:
        for relative, size in oversized:
            print(
                f"error: {relative} is {size / 1048576:.1f} MiB, over the "
                f"{MAX_ASSET_BYTES / 1048576:.0f} MiB Cloudflare asset limit",
                file=sys.stderr,
            )
        return 1

    # index.html and 404.html anchor the routing configured in wrangler.jsonc.
    for required in ("index.html", "404.html"):
        if not (out_dir / required).is_file():
            print(f"error: {required} missing from build output", file=sys.stderr)
            return 1

    print(f"Built {len(files)} files ({total_bytes / 1048576:.1f} MiB) into {out_dir}")
    return 0


def main() -> int:
    parser = argparse.ArgumentParser(description="Build the static site into ./dist")
    parser.add_argument(
        "--out",
        default=str(ROOT / "dist"),
        help="output directory (default: ./dist)",
    )
    args = parser.parse_args()
    return build(ROOT, Path(args.out).resolve())


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/build_site.js
#!/usr/bin/env node
/**
 * PipeFish Labs - Static Site Build (Node.js)
 *
 * Assembles the publicly servable website into ./dist so that it can be uploaded
 * as Cloudflare Workers static assets.
 *
 * Pure Node.js with zero external dependencies.
 */

import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const ROOT = path.resolve(__dirname, '..');

const ALLOWED_SUFFIXES = new Set([
  '.html',
  '.css',
  '.js',
  '.mjs',
  '.png',
  '.jpg',
  '.jpeg',
  '.gif',
  '.svg',
  '.webp',
  '.avif',
  '.ico',
  '.mp4',
  '.webm',
  '.woff',
  '.woff2',
  '.txt',
  '.xml',
  '.webmanifest',
  '.pdf',
]);

const EXCLUDED_DIRS = new Set([
  '.git',
  '.github',
  '.githooks',
  '.venv',
  '.wrangler',
  '__pycache__',
  'node_modules',
  'dist',
  'ansible',
  'bots',
  'charts',
  'db',
  'docs',
  'examples',
  'infra',
  'linkedin',
  'mcp',
  'postman',
  'pipefish_labs.egg-info',
  'sales',
  'scripts',
  'security',
  'src',
  'terraform',
  'tests',
]);

const EXTRA_FILES = [
  'CNAME',
  '_headers',
  '_redirects',
  'admin/config.yml',
  'sdk/openapi.yaml',
  'sdk/asyncapi.yaml',
  'site.webmanifest',
  '.well-known/security.txt',
];

const MAX_ASSET_BYTES = 25 * 1024 * 1024;

const EXCLUDED_FILES = new Set([
  'briefing-video-poster.png',
  'briefing-video-poster.jpg',
]);

function isPublishable(filePath) {
  const ext = path.extname(filePath).toLowerCase();
  const basename = path.basename(filePath);
  return ALLOWED_SUFFIXES.has(ext) && !EXCLUDED_FILES.has(basename);
}

function walkDir(currentDir, relativePrefix, fileList) {
  const entries = fs.readdirSync(currentDir, { withFileTypes: true });

  for (const entry of entries) {
    const fullPath = path.join(currentDir, entry.name);
    const relPath = relativePrefix ? path.join(relativePrefix, entry.name) : entry.name;

    if (entry.isDirectory()) {
      if (EXCLUDED_DIRS.has(entry.name)) {
        continue;
      }
      walkDir(fullPath, relPath, fileList);
    } else if (entry.isFile()) {
      if (isPublishable(fullPath)) {
        fileList.push(relPath);
      }
    }
  }
}

export function build(rootDir = ROOT, outDir = path.resolve(ROOT, 'dist')) {
  if (outDir === rootDir || outDir.startsWith(rootDir + path.sep) && path.relative(rootDir, outDir) === '') {
    console.error(`error: refusing to use ${outDir} as the output directory`);
    return 1;
  }

  const fileList = [];
  walkDir(rootDir, '', fileList);

  const seen = new Set(fileList.map((f) => path.normalize(f)));
  const missing = [];

  for (const extra of EXTRA_FILES) {
    const fullPath = path.join(rootDir, extra);
    if (!fs.existsSync(fullPath)) {
      missing.push(extra);
      continue;
    }
    const norm = path.normalize(extra);
    if (!seen.has(norm)) {
      fileList.push(extra);
      seen.add(norm);
    }
  }

  if (missing.length > 0) {
    for (const name of missing) {
      console.error(`error: required file is missing: ${name}`);
    }
    return 1;
  }

  if (fileList.length === 0) {
    console.error('error: no publishable files found');
    return 1;
  }

  if (fs.existsSync(outDir)) {
    fs.rmSync(outDir, { recursive: true, force: true });
  }
  fs.mkdirSync(outDir, { recursive: true });

  let totalBytes = 0;
  const oversized = [];

  // Sort files for deterministic builds
  fileList.sort();

  for (const relative of fileList) {
    const source = path.join(rootDir, relative);
    const stat = fs.statSync(source);
    totalBytes += stat.size;

    if (stat.size > MAX_ASSET_BYTES) {
      oversized.push([relative, stat.size]);
    }

    const destination = path.join(outDir, relative);
    const destDir = path.dirname(destination);
    fs.mkdirSync(destDir, { recursive: true });
    fs.copyFileSync(source, destination);
  }

  if (oversized.length > 0) {
    for (const [rel, size] of oversized) {
      console.error(
        `error: ${rel} is ${(size / 1048576).toFixed(1)} MiB, over the ${(MAX_ASSET_BYTES / 1048576).toFixed(0)} MiB Cloudflare asset limit`
      );
    }
    return 1;
  }

  for (const required of ['index.html', '404.html']) {
    if (!fs.existsSync(path.join(outDir, required))) {
      console.error(`error: ${required} missing from build output`);
      return 1;
    }
  }

  console.log(`Built ${fileList.length} files (${(totalBytes / 1048576).toFixed(1)} MiB) into ${outDir}`);
  return 0;
}

// Direct CLI execution
if (process.argv[1] && path.resolve(process.argv[1]) === __filename) {
  let outArg = path.resolve(ROOT, 'dist');
  const outIdx = process.argv.indexOf('--out');
  if (outIdx !== -1 && process.argv[outIdx + 1]) {
    outArg = path.resolve(process.argv[outIdx + 1]);
  }
  const code = build(ROOT, outArg);
  process.exit(code);
}
